### DISCUSSION TITLE
$screen helper

### DISCUSSION THREAD

**Discussion by SimoTod** [alpine-collective/alpine-magic-helpers#37] [Ideas] [answered]
Let's chat about this.
$screen helper which returns true if the screen size is equal to or bigger than a given breakpoint.

Discussion point: Should we hardcode external frameworks settings inside the plugin, ask the user to specify the breakpoints somewhere, or just give a sensible default (documented) and allow the user to override it easily.

My personal view is that people tomorrow will question why X, Y, Z are supported but the new A framework is not, so I'm not a fan of giving those configurations here. On the other side, I'm not sure that asking people to define breakpoint by themselves would be a good idea so I'm a bit torn.

**Comment by MuzafferDede** [comment, marked as answer]
I would go with defining breakpoints by user instead of writing in the code. I am not sure what path would be the best to define these breakpoints only once  and let $screen helper to initialize it. 

Perhaps we could at least provide tailwindcss grid as default and a way to provide override? 

@KevinBatdorf for your suggestion on callback, what would be the use case? Even there is one, can we just use `$watch` helper for that?

My add on to this helper would be providing `touch` detection, maybe `portrait` and `landscape` mode as well. 

I used the have this idea of having a way to set Alpine related configurations to make almost anything configurable, but i think this would be a better approach for V3. (Not sure if there is still V3 roadmap tho)

**Reply by KevinBatdorf** [reply]
Since Alpine is built as a companion to Livewire, and Tailwind is a big part of the PHP community and ecosystem, I think that's a good argument for having Tailwind in there as a default (Alpine docs use Tailwind throughout too). I also use it personally so it gets my vote as a default.

We can add an optional `$breakpoint()` helper that can override the defaults easily enough. Maybe include the bulma or bootstrap override in the docs as an example.

What's the use case for a callback? Not totally confident about it. It's something we can add later if it makes sense.  OK to hold on that for now.

**Reply by MuzafferDede** [reply]
Yup, tailwind is everywhere now. Making it default is a big plus.

Help me to understand, if we go with the `$breakpoint` helper way, how are we going to define them for each components uses `$screen` helper? Since `$breakpoint` will be only available within the defined component, there is no re-usability, or am i wrong? 
I am not sure if we can apply it on javascript level directly to Alpine and in `$screen` helper check if there is any breakpoints are defined, if so, use them, if not, use default as tailwind grid.

I am heavily using vue-screen and never needed a callback, also most of the vue-screen functionalities. The way i implemented in alpine is clean and not too complex as Alpine itself.

**Reply by SimoTod** [reply]
I'm happy with using Tailwind as a default. 

About overriding it, should we expose a hook on the global namespace (window) so we can have something like
```html
<script src="https://cdn.jsdelivr.net/gh/alpine-collective/alpine-magic-helpers@0.4.x/dist/index.min.js"></script>
<script>AlpineBreakpoints.override({
  'xs': 0,
  'md': 600,
  ...
})
<script type="module" src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.x.x/dist/alpine.min.js"></script>

**Reply by KevinBatdorf** [reply]
That works for me. The global override makes more sense.

I doubt there is any need to override them on a "per component" basis (ie `$breakpoint()`), but we can always add that sort of override later if someone has a good case for it.

**Reply by MuzafferDede** [reply]
@SimoTod how about something re-useable as ConfigBus ? i think there might be other cases we might need some sort of storage for global definitions. By extending ConfigBus, we can implement and extend any helper needs.

**Reply by SimoTod** [reply]
Is it a third party library? 

I was thinking we can just export a config object from the screen helper and import/export is as it is in index php so it gets exposed on the global namespace and it works consistently for both the full bundle and the standalone helper, but if you have a different strategy to achieve the same result, we can have a look. 👍

**Reply by MuzafferDede** [reply]
No, it is not any library but just a concept of an idea. It's mostly as how you described. I was thinking more like utility like global object called `ConfigBus` where every helper checks if it's exists and extend it as it need. If not exists then create one with the configuration values it needs. But looking at the current helpers, we might not need right now such a thing. But in case other helpers need to share configuration between, we could consider this concept.

**Reply by SimoTod** [reply]
In that case, yeah a structured object will work as long as it's exposed on the global namespace (window)

**Reply by KevinBatdorf** [reply]
Like `window.AlpineConfig` and `AlpineConfig.breakpoints()` or should we use `AlpineMagicHelpers` instead?

**Reply by MuzafferDede** [reply]
or can we assign as `window.Alpine.config = {...}`?  maybe in future alpine would merge this into core so we wouldn't have any breaking changes.

**Reply by SimoTod** [reply]
I vote for AlpineMagicHelpers.Config.breakpoints
AlpineConfig sounds like it's an official feature

**Reply by MuzafferDede** [reply]
Yeah, that will work i guess. Also we could use AlpineMagicHelpers as a global utility helper for any type of hooks.

**Reply by SimoTod** [reply]
I think it's safer if we don't use the Alpine namespace. The name could conflict in the future and we don't own the library

**Reply by KevinBatdorf** [reply]
`AlpineMagicHelpers.Config.breakpoints` sounds good to me.

**Reply by MuzafferDede** [reply]
I think we have a solid idea for the config part. @KevinBatdorf @SimoTod is it ok for you to do the AlpineMagicHelpers.Config with another PR then i re-factor $screen helper afterwards?

**Reply by KevinBatdorf** [reply]
I can make a PR. Not sure the best approach so I'll just implement what comes to mind first and we can discuss it.

**Comment by KevinBatdorf** [comment]
Another topic, should the debounce should be an option you can pass in? `$screen('xl', {debounce: 0})` maybe? I don't necessarily like how it reads, but I can see it being useful.

**Reply by MuzafferDede** [reply]
My thoughts on debounce is `$screen` helper is a power hungry task because it updates element using $el.updateElements($el), which means everything on the component will be re-calculated. That's why i use debounce to prevent it. Giving users this as functionality might cause issues on large trees.

**Reply by SimoTod** [reply]
I think it's safer to just debounce for now as in the initial PR and add it later if we can find a use case where we want to configure it.

**Reply by KevinBatdorf** [reply]
I'm ok with that. Should we match Alpine's `250` default on inputs or is that too slow?

**Reply by MuzafferDede** [reply]
i think that might be a bit too slow, i use tailwindcss default duration which is 150.

**Comment by SimoTod** [comment]
I'm going to open a new discussion point for the second iteration of the `$screen` helper.
Should we support the screen orientation,`$screen('portrait')` and `$screen('landscape')`?
If yes, should it be the same helper or a new one? We need to consider that the breakpoint will need a listener on the resize event while an orientation change won't need that one but it will need to listen to a different event.
Also, should we implement it using height and width or just using window.matchMedia? The advantage of the latter is that the browser will work consistently with css for edge cases (for example when height and width are exactly the same).

**Reply by KevinBatdorf** [reply]
Can you explain the issue with the event listeners a bit more? Does this have to do with some events not firing?

> The advantage of the latter is that the browser will work consistently with css for edge cases
Sounds like the most future-proof approach.

**Reply by HugoDF** [reply]
this other event? https://developer.mozilla.org/en-US/docs/Web/API/Window/orientationchange_event

**Reply by KevinBatdorf** [reply]
Does a resize event not fire at the same time, or are we not sure? Could we just update the code to this:

```js
window.addEventListener('resize', debouncedScreensUpdate())
window.addEventListener('orientationchange', debouncedScreensUpdate())
```

... and not worry about it? The debounce would take care of devices that fire both, right?

**Reply by SimoTod** [reply]
muzafferdede checked and he said that rotating your phone does not trigger a resize event (TBC) so components won't update.
We'll need to listen for the event Hugo posted if that's the case.

**Reply by HugoDF** [reply]
`$screen` would be nice to have with both orientation, touch etc

if we decided to split them I guess we would do something like `$breakpoint('lg')` and `$is('touch')` or `$detect('portrait')`

The thing with portrait/landscape is that you can deal with it in CSS so dunno how much value it'll add to also be able to access them in JS

**Reply by MuzafferDede** [reply]
@SimoTod 
> Should we support the screen orientation,$screen('portrait') and $screen('landscape')?

@HugoDF 
> $screen would be nice to have with both orientation, touch etc

My initial thoughts was include anything related to screen to `$screen` helper. Didn't really want to increase the amount of the helpers to cause learning edge. `jest` and @KevinBatdorf's battle made me give up  for now 😄 

@HugoDF 
>The thing with portrait/landscape is that you can deal with it in CSS so dunno how much value it'll add to also be able to access them in JS

TBH, `touch`, and `orientation` support were add ons after initial thought. They are just **why not**s 

@SimoTod 
> muzafferdede checked and he said that rotating your phone does not trigger a resize event (TBC) so components won't update.

@SimoTod i did check my comments if i said such a thing but i guess i didn't. Because i did check and actually even implemented a test as well for the `orientationchange` event which triggers `resize` event someway. So `orientationchange` = `resize` on mobile as well. 

The things i couldn't do was tests  on jest. Because there is no  `orientationchange` support. I tried to mock and couldn't figure out. Sorry if i made any miss leading.

**Reply by SimoTod** [reply]
@muzafferdede you did here -> https://github.com/alpine-collective/alpine-magic-helpers/pull/35#discussion_r528333282
After I asked if someone could check. 

Not sure what game you are playing. :/

**Reply by SimoTod** [reply]
Anyway, topic closed, we don't need a listener on orientation change which is good.

**Reply by MuzafferDede** [reply]
> @muzafferdede you did here -> [#35 (comment)](https://github.com/alpine-collective/alpine-magic-helpers/pull/35#discussion_r528333282)
> After I asked if someone could check.
> 
> Not sure what game you are playing. :/

Oh sorry, my bad, my bad. Stupid pass me! 😢 It's probably me just discovering things at that moment :) After that I actually tested it by console.log() in both event to see if I got 2 log messages. Which was so, then I did some more test and figure out mobile does trigger resize event. There was too much going on at that time. My bad.

BTW, I only play single game, DOTA 2 but I code more, which is more fun. You know 🙄
